### ISSUE TITLE
The loading order changes the index offset

### ISSUE THREAD

**Issue by reemplazable** [pat/thinking-sphinx#992]
Just happened to me:
Loading the indexed dinamically (they are real time) makes them take different offset between the application and the rake ts:regenerate :

```
document_id_for_key 647 key: 22 indices_count: 29 offset: 9
document_id_for_key 643 key: 22 indices_count: 29 offset: 5
```

The index offset should be the same regardless on the loading order.

```
  describe '#next_offset' do
    it "returns the same offset regardless of loading order" do
      config.next_offset('a').should == 0
      config.next_offset('b').should == 1
      ThinkingSphinx::Configuration.reset
      config.next_offset('b').should == 1
      config.next_offset('a').should == 0
    end
  end
```

**Comment by pat** [comment]
Can you provide some more details on how you're loading indices dynamically? Are you doing anything with Thinking Sphinx's internals? Or just using TS in a rather standard way?

**Comment by reemplazable** [comment]
It is used in a standar way, I defined various index that have the same content for search.
```
module QuestionIndex
  def self.create(question_class)
    ThinkingSphinx::Index.define question_class.name.underscore.to_sym, with: :real_time do
      indexes :index_title, as: :title, type: :string
      # more code here with properties.
    end
  end
end
```
I use a little bit of metaprograming because index are from the same class heritance.
Then, in each engine I've got a call to create the index:
`QuestionIndex.create OneType::Question`
`QuestionIndex.create AnotherType::Question`

So, when the application starts it asigns an offset to each index as I can see in the code. The problem is, the index loading order (sounds like class/file ruby loading order is changing) changes from the application start to be served and the application to be used by rake to ts:regenerate. So, when in the indexes, as You can see from the trace I put before, the same class index has different Id in sphinx and creates duplicated records.

So, next_offset should give the same offset for each class index independent of the loading order.

From my point of view (trying not to contaminate with a solution):
Maybe the offset should be a CRC32 of the index name instead of a next in a count. So every time the offset is calculated is always the same one and you end up in a test like this

```
  describe '#next_offset' do
    it "returns the same offset regardless of loading order" do
      config.next_offset('a') = a_offset
      config.next_offset('b') = b_offset
      ThinkingSphinx::Configuration.reset
      config.next_offset('b').should == b_offset
      config.next_offset('a').should == a_offset
    end
  end
```

**Comment by pat** [comment]
The tricky part about making `next_offset` aware of which indices exist is that it's used as part of the index definition flow.

TS expects indices to be defined in `app/indices` (whether in the app itself, or in Rails engines that are present), and loads these in a consistent order: https://github.com/pat/thinking-sphinx/blob/develop/lib/thinking_sphinx/configuration.rb#L82 - so, if it's possible, I'd recommend putting each of these `QuestionIndex.create` calls into separate files in the appropriate `app/indices` directories. This will make sure that TS always loads them when it's determining all the index definitions are present, and thus, the offsets should always be the same.

**Comment by reemplazable** [comment]
Maybe I did not explain myself clear, that calls where in each app/indices of each engine.

**Comment by pat** [comment]
Ah, okay. Hmm. Have you tracked down the situation where the offsets change from the original values? Is it with all creates/updates (via the ActiveRecord callbacks)? Or only in specific cases?

**Comment by reemplazable** [comment]
The difference is between using the app through the Passenger server and a rake ts:regenerate

**Comment by pat** [comment]
And which version of TS are you using?

**Comment by reemplazable** [comment]
Now, lastets (3.2.0), but I started with the 3.1.1

**Comment by pat** [comment]
And you're still seeing the issue with 3.2.0? There was a thread-safety patch in that release related to the index preloading (98a51374), so I'm wondering if the issue is related.

**Comment by reemplazable** [comment]
Yes, there is no difference between both versions. :-(

**Comment by pat** [comment]
I realise it's been a while, but I've been pondering this problem and trying to think through what the cause could be. The loading of indices (and thus, the calculation of offsets) is wrapped in a mutex and as far as I can tell is thread-safe.

The only possibility for difference that I've been able to spot is if the order of engines loaded in your app differs - which would be very odd! Surely gems (and thus engines) are loaded in a consistent order before threads are involved - but I've just pushed a commit to the `develop` branch which should account for this just in case. You're welcome to give that a shot if you'd like!

Are you finding this behaviour is consistent via Passenger? i.e. does it apply to every single record, every single time, and with the same miscalculated offset?

I'm yet to think of a different approach to calculate offsets. As you've suggested, CRC32 could be a way forward, but it is possible to have conflicts with CRC32 generation, so I don't feel it's reliable. It'd also end up with massive offsets, which means hitting the upper bounds of large integers would become far more frequent. But besides that, we need to know about what indices are present to get the name/offset pairs consistent, and that means we're back to the original problem - at least, what _seems_ to be the problem - in that the indices aren't being loaded in the same order in Passenger compared to rake.
